services/wallet/thirdparty/coingecko/client.go

```diff
@@ -96,13 +101,35 @@ func (c *Client) DoQuery(url string) (*http.Response, error) {
 	return resp, nil
 }
 
+func mapTokensToSymbols(tokens []GeckoToken, tokenMap map[string]GeckoToken) {
+	for _, token := range tokens {
+		if id, ok := coinGeckoMapping[strings.ToUpper(token.Symbol)]; ok {
+			if id != token.ID {
+				continue
+			}
+		}
+		tokenMap[strings.ToUpper(token.Symbol)] = token
+	}
+}
+
 func (c *Client) getTokens() (map[string]GeckoToken, error) {
+	fetching := (&c.isTokenFetchPending).Load()
+
+	if fetching {
+		c.fetchTokensMutex.Lock()
+		defer c.fetchTokensMutex.Unlock()
+	}
+
 	if len(c.tokens) > 0 {
 		return c.tokens, nil
 	}
 
-	url := fmt.Sprintf("%scoins/list", baseURL)
-	resp, err := c.DoQuery(url)
+	c.fetchTokensMutex.Lock()
```

Comment:
There's something I don't understand here. If you enter the `fetching` condition above, AFAIK the `defer`ed `Unlock()` will be called at the end of the function. If for whatever reason `len(c.tokens) == 0` and you call `Lock()` here again, won't that be a deadlock?